### PR TITLE
Sort question bank / question list questions by most recently modified

### PR DESCRIPTION
**Pull request by clouser2**
### Description

This makes use of the sorting mechanism added for the program list and applies it to both the question bank and question list. Additionally, a "sorted by" chip was added to the question / program list (screenshots below).

Sorting is done client-side per the rationale in https://github.com/civiform/civiform/issues/3029#issuecomment-1238371242. This approach was generalized and moved to `sorting.ts`. `sorting.ts` registers all of the elements on which it will sort rather than providing a general helper utility since we don't currently have the ability to use JavaScript modules (https://github.com/civiform/civiform/issues/3029#issuecomment-1238399530).

Finally, the "placeholder" mechanism was removed since it added complexity and wasn't actually working for the program list anyway.

### Screenshots

Question list before:
Screen Shot 2022-09-07 at 12 49 27 PM

Question list after:
Screen Shot 2022-09-07 at 12 46 17 PM

Program list before:
Screen Shot 2022-09-07 at 12 49 21 PM

Program list after:
Screen Shot 2022-09-07 at 12 46 12 PM

## Release notes:

Sorting the question list and questions shown in the program builder by the most recently updated question.

### Checklist

- [X] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [X] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)

#3029

### CODE REVIEW

File: browser-test/src/civiform_admin_question_list.test.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import {
+  AdminPrograms,
+  AdminQuestions,
+  createTestContext,
+  loginAsAdmin,
+} from './support'
+
+describe('Most recently updated question is at top of list.', () => {
```
Comment (clouser2) [inline browser-test/src/civiform_admin_question_list.test.ts:8]: Note to the reviewer: This test was modeled off of the corresponding admin_program_list.test.ts (https://github.com/civiform/civiform/blob/08868c0bc54327e3ed4289c9827ba5c471f109e8/browser-test/src/admin_program_list.test.ts)

File: browser-test/src/civiform_admin_question_list.test.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    const {page, adminPrograms, adminQuestions} = ctx
+
+    await loginAsAdmin(page)
+
+    const questionOne = 'question list test question one'
+    const questionTwo = 'question list test question two'
+    await adminQuestions.addNameQuestion({questionName: questionOne})
+    await adminQuestions.addNameQuestion({questionName: questionTwo})
+
+    // Note: CI tests already have test questions
```
Comment (nbeloglazov) [inline browser-test/src/civiform_admin_question_list.test.ts:20]: Is that true? Each test file starts with empty database. And if you have emptry database you can assume only your questions are present and simplify assertions below.

File: browser-test/src/support/admin_programs.ts [hunk 3]
```diff
@@ -253,6 +253,14 @@ export class AdminPrograms {
     }
   }
 
+  async questionBankNames(programName: string) {
```
Comment (nbeloglazov) [inline browser-test/src/support/admin_programs.ts:256]: Add return type `string[]` given that it's public function.

File: browser-test/src/support/admin_questions.ts [hunk 4]
```diff
@@ -244,6 +244,14 @@ export class AdminQuestions {
     )
   }
 
+  async questionNames() {
```
Comment (nbeloglazov) [inline browser-test/src/support/admin_questions.ts:247]: Same here.

File: server/app/assets/javascripts/sorting.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  // Question bank.
+  const questionBankParent = document.getElementById('question-bank-questions')
+  if (questionBankParent) {
+    ElementSorter.sortElementsOnLoad({
+      parent: questionBankParent,
+      compareFunc: lastUpdatedAndNameComparator,
+    })
+  }
```
Comment (nbeloglazov) [inline server/app/assets/javascripts/sorting.ts:89]: Looks like all these 3 blocks are mostly the same except for the selector. Maybe do process them in a loop?

```
for (const selector of ['.cf-admin-question-list', ...]) {
  const el = document.querySelector(selector);
  if (!el) continue;
  ...
}
```

File: server/app/assets/javascripts/sorting.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  }) {
+    const elements = Array.from(
+      parent.querySelectorAll(ElementSorter.sortableElementSelector),
+    )
+    try {
+      ElementSorter.sortCards(parent, elements, compareFunc)
+    } finally {
+      // Make sure to always show the element, even
+      // if there was an error while sorting.
+      parent.classList.remove('invisible')
```
Comment (nbeloglazov) [inline server/app/assets/javascripts/sorting.ts:22]: That makes me a bit nervous that we rely on JS to make element visible. So if JS is slow to load user will see empty page. I think ideally we'd want to sort server-side and add ability to re-sort client-side if necessary. But it's more work to maintain both server-side and client-side sorting... Do we expect other client-side sorting options to be added soon?

**Review by nbeloglazov — CHANGES_REQUESTED** [review]

**Comment by clouser2** [comment]
This surfaced some existing issues where we weren't properly awaiting scripts after clicking a button which would trigger a navigation. While I rework things, I'll take it out of review.

**Review by clouser2 — COMMENTED** [review]
Replying to comments for posterity. This PR is being dropped in favor of https://github.com/civiform/civiform/pull/3431 (where all feedback has been incorporated). We'll now do sorting server-side until we have a real need for client-side sorting.

**Pull request closed without merging**